test/test.js

```diff
@@ -4997,6 +4997,150 @@ exports['flatMap - map to Stream of Array'] = function (test) {
+            test.same(xs, [2, 4, 6, 8]);
+            test.done();
+        });
+    },
+    'noValueOnError': noValueOnErrorTest(_.ap(_.of(1))),
+    'ArrayStream': function (test) {
+        test.expect(1);
+        var f = _.of(function (x) {
+            return _(function (push, next) {
+                setTimeout(function () {
+                    push(null, x * 2);
+                    push(null, _.nil);
+                }, 10);
+            });
+        });
+        _([1, 2, 3, 4]).ap(f).merge().toArray(function (xs) {
+            test.same(xs, [2, 4, 6, 8]);
+        });
+        this.clock.tick(20);
+        test.done();
+    },
+    'GeneratorStream': function (test) {
+        test.expect(1);
+        var f = _.of(function (x) {
+            return _(function (push, next) {
+                push(null, x * 2);
+                push(null, _.nil);
+            });
+        });
+        var s = _(function (push, next) {
+            push(null, 1);
+            push(null, 2);
+            push(null, 3);
+            push(null, 4);
+            push(null, _.nil);
+        });
+        s.ap(f).merge().toArray(function (xs) {
+            test.same(xs, [2, 4, 6, 8]);
+            test.done();
+        });
+    },
+    'map to Stream of Array': function (test) {
+        test.expect(1);
+        var f = _.of(function (x) {
+            return _([[x]]);
+        });
+        var s = _([1, 2, 3, 4]).ap(f).merge().toArray(function (xs) {
+            test.same(xs, [[1], [2], [3], [4]]);
+            test.done();
+        });
+    },
+    'reflect timing of value and function arrival': function (test) {
+        test.expect(4);
+        var f = _(function (push, next) {
+            setTimeout(function () {
+                push(null, function (x) { return 'g1(' + x + ')'; });
+            }, 20);
+            setTimeout(function () {
+                push(null, function (x) { return 'g2(' + x + ')'; });
+                push(null, _.nil);
+            }, 60);
+        });
+        var s = _(function (push, next) {
+            setTimeout(function () {
+                push(null, 1);
+            }, 10);
+            setTimeout(function () {
+                push(null, 2);
+            }, 50);
+            setTimeout(function () {
+                push(null, 3);
+                push(null, _.nil);
+            }, 70);
+        });
+
+        var results = [];
+        s.ap(f).each(function (x) {
+            results.push(x);
+        });
+        this.clock.tick(20);
+        test.same(results, ['g1(1)']);
+        this.clock.tick(30);
+        test.same(results, ['g1(1)', 'g1(2)']);
+        this.clock.tick(10);
+        test.same(results, ['g1(1)', 'g1(2)', 'g2(2)']);
+        this.clock.tick(10);
+        test.same(results, ['g1(1)', 'g1(2)', 'g2(2)', 'g2(3)']);
+        test.done();
+    },
+    // v.ap(u.ap(a.map(f => g => x => f(g(x))))) is equivalent to v.ap(u).ap(a) (composition)
+    'composition': {
```

Comment:
I think it's important that `u` and `v` and `a` are defined the same way, so you should be defining them in the same test. Having different tests like this means they can diverge in the future.

You should, have one test that

1. Defines `u`, `v`, and `a` once.
2. Builds `v.fork().ap(u.fork().ap(a.fork().map(f => g => x => f(g(x)))))` and `a.fork().ap(u.fork()).ap(a.fork())`
3. Consume both of them.
4. Verifies that both output `abc`.